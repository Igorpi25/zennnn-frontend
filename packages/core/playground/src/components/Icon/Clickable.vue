<template>
  <Icon
    class="mr-6"
    @click="onClick"
  >
    {{ ziEdit }}
  </Icon>
  <Icon
    tag="button"
    role="button"
    large
    class="text-blue-500 hover:text-blue-600 focus:text-blue-600"
    @click="onClick"
  >
    {{ ziEdit }}
  </Icon>
</template>

<script setup>
import { ziEdit } from '@zennnn/icons'

function onClick () {
  console.log('Clicked!')
}
</script>
