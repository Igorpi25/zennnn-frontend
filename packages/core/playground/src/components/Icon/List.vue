<template>
  <div class="flex flex-wrap">
    <div
      v-for="(icon, name) of icons"
      :key="name"
      class="w-1/2 sm:w-40 sm:px-4 mt-2 mb-3"
    >
      <div class="text-gray-400 text-sm truncate pb-1">{{ name.replace('zi', '') }}</div>
      <Icon>
        {{ icon }}
      </Icon>
    </div>
  </div>
</template>

<script setup>
import * as icons from '@zennnn/icons'
</script>
