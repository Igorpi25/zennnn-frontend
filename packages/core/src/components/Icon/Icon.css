.icon {
  display: inline-flex;
  font-feature-settings: 'liga';
  justify-content: center;
  letter-spacing: normal;
  line-height: 1;
  text-indent: 0;
  vertical-align: middle;
  user-select: none;
}
.icon svg {
  text-indent: 0;
  letter-spacing: normal;
  line-height: 1;
  user-select: none;
  fill: currentColor;
  vertical-align: middle;
  width: 1em;
  height: 1em;
}
.icon--link {
  @apply cursor-pointer rounded;
  @apply focus:outline-none focus:ring focus:ring-blue-400;
}
.icon--link:focus:not(:focus-visible) {
  box-shadow: none;
  outline: 0;
}
