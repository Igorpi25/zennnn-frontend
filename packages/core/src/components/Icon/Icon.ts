import { h, computed, defineComponent } from 'vue'
import { convertToUnit } from 'vue-supp'

const SIZE_CLASS_MAP = {
  small: '18px',
  base: '24px',
  large: '32px',
}

export default defineComponent({
  name: 'Icon',

  props: {
    tag: {
      type: String,
      default: 'i',
    },
    size: {
      type: [Number, String],
      default: undefined,
    },
    small: Boolean,
    large: Boolean,
    base: {
      type: Boolean,
      default: true,
    },
    left: Boolean,
    right: Boolean,
  },

  setup(props, { attrs, slots }) {
    const fontSize = computed((): string | undefined => {
      return props.size
        ? convertToUnit(props.size)
        : props.small
        ? SIZE_CLASS_MAP.small
        : props.large
        ? SIZE_CLASS_MAP.large
        : props.base
        ? SIZE_CLASS_MAP.base
        : undefined
    })

    const genSvg = () => {
      const node = slots.default?.() ?? []
      const children = node[0] && node[0].children
      const icon = typeof children === 'string' ? children : undefined

      if (!icon) return undefined

      const svgData: Record<string, any> = {
        xmlns: 'http://www.w3.org/2000/svg',
        viewBox: '0 0 24 24',
        height: fontSize.value || '24',
        width: fontSize.value || '24',
        role: 'img',
        'aria-hidden': true,
      }
      if (fontSize.value) {
        svgData.style = {
          fontSize: fontSize.value,
          height: fontSize.value,
          width: fontSize.value,
        }
      }

      return h(
        'svg',
        svgData,
        h(
          'g',
          {
            fill: 'currentColor',
          },
          h('path', {
            d: icon,
          })
        )
      )
    }

    const clickable = attrs.onClick

    return () => {
      const tag = clickable ? 'button' : props.tag
      return h(
        tag,
        {
          class: {
            icon: true,
            'icon--left': props.left,
            'icon--right': props.right,
            'icon--link': clickable,
          },
          role: clickable ? 'button' : undefined,
          'aria-hidden': !clickable,
        },
        genSvg()
      )
    }
  },
})
