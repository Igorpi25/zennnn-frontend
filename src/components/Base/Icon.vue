<script>
import { convertToUnit } from '@/util/helpers'

export default {
  name: 'Icon',
  props: {
    iconName: {
      type: String,
      default: '',
    },
    title: {
      type: [String, Object],
      default: '',
    },
    size: {
      type: [String, Number],
      default: 18,
    },
    color: {
      type: String,
      default: 'currentColor',
    },
  },
  computed: {
    path () {
      return `@/assets/icons/${this.iconName}.svg`
    },
  },
  render (createElement) {
    const h = createElement
    const attrs = {
      xmlns: 'http://www.w3.org/2000/svg',
      width: this.size,
      height: this.size,
      viewBox: '0 0 24 24',
    }
    const size = convertToUnit(this.size)
    if (this.iconName) {
      return h('i', {}, [
        h('img', {
          attrs: {
            src: this.path,
          },
        }),
      ])
    } else if (this.$slots.default[0].text) {
      return h('svg', {
        attrs,
        style: {
          width: size,
          height: size,
          'font-size': size,
          color: this.color,
        },
      },
      [h('title',
        this.title,
      ),
      [h('g', {
        attrs: {
          fill: this.color,
        },
      },
      [h('path', {
        attrs: {
          d: this.$slots.default[0].text,
        },
      })])]])
    }
    return null
    // else {
    //   const comp = this.$slots.default[0]
    //   comp.componentOptions.propsData = {
    //     width: attrs.width,
    //     height: attrs.height,
    //     styles: {
    //       width: size,
    //       height: size,
    //       'font-size': size,
    //     },
    //   }
    //   return comp
    // }
  },
}
</script>
