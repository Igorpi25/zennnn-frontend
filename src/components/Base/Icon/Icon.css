@layer components {
  .icon {
    align-items: center;
    display: inline-flex;
    font-feature-settings: 'liga';
    justify-content: center;
    letter-spacing: normal;
    line-height: 1;
    text-indent: 0;
    vertical-align: middle;
    user-select: none;
  }
  .icon svg {
    text-indent: 0;
    letter-spacing: normal;
    line-height: 1;
    user-select: none;
    fill: currentColor;
    vertical-align: middle;
    width: 1em;
    height: 1em;
  }
}
