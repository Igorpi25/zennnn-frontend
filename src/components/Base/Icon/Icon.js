import { h, computed } from 'vue'

import { convertToUnit } from '../../../utils/convertToUnit'

import './Icon.css'

const SIZE_CLASS_MAP = {
  small: '18px',
  base: '24px',
  large: '32px',
}

export default {
  name: 'Icon',

  props: {
    tag: {
      type: String,
      default: 'i',
    },
    size: {
      type: [Number, String],
      default: undefined,
    },
    small: Boolean,
    large: Boolean,
    base: {
      type: Boolean,
      default: true,
    },
  },

  setup (props, { attrs, slots }) {
    const tag = computed(() => props.tag || 'i')

    const fontSize = computed(() => {
      return props.size
        ? convertToUnit(props.size)
        : props.small
          ? SIZE_CLASS_MAP.small
          : props.large
            ? SIZE_CLASS_MAP.large
            : props.base
              ? SIZE_CLASS_MAP.base
              : undefined
    })

    const genSvg = () => {
      const node = slots.default ? slots.default() : []
      const children = node[0] && node[0].children
      const icon = typeof children === 'string' ? children : undefined

      if (!icon) return undefined

      const svgData = {
        xmlns: 'http://www.w3.org/2000/svg',
        viewBox: '0 0 24 24',
        height: fontSize.value || '24',
        width: fontSize.value || '24',
        role: 'img',
        'aria-hidden': true,
      }
      if (fontSize.value) {
        svgData.style = {
          fontSize: fontSize.value,
          height: fontSize.value,
          width: fontSize.value,
        }
      }

      return h('svg', svgData, h('g', {
        fill: 'currentColor',
      }, h('path', {
        d: icon,
      })))
    }

    const classes = { icon: true }

    const clickable = attrs.onClick

    return () => {
      return h(tag.value, {
        class: classes,
        role: clickable ? 'button' : undefined,
      }, genSvg())
    }
  },
}
