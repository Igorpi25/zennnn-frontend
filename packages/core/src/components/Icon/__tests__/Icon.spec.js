import { mount } from '@vue/test-utils'
import Icon from '../Icon'

describe('Icon.js', () => {
  let mountFunction

  beforeEach(() => {
    mountFunction = (options = {}) => {
      return mount(Icon, {
        slots: {
          default: () => 'M18 12a6 6 0 11-12 0 6 6 0 0112 0z',
        },
        ...options,
      })
    }
  })

  // Render

  it('should render component and match snapshot', () => {
    const wrapper = mountFunction()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render specified tag when using tag prop and match snapshot', () => {
    const wrapper = mountFunction({
      props: { tag: 'span' },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render "button" tag on attrs.onClick exists and match snapshot', () => {
    const wrapper = mountFunction({
      props: { tag: 'span' },
      attrs: {
        onClick: () => {},
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should set size from helper prop and match snapshot', () => {
    const iconFactory = size => mountFunction({
      props: { [size]: true },
    })

    const small = iconFactory('small')
    expect(small.html()).toMatchSnapshot()

    const medium = iconFactory('base')
    expect(medium.html()).toMatchSnapshot()

    const large = iconFactory('large')
    expect(large.html()).toMatchSnapshot()
  })

  it('should have correct size style on size prop and match snapshot', () => {
    const wrapper = mountFunction({
      props: {
        size: 20,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should not set size style on base prop false and match snapshot', () => {
    const wrapper = mountFunction({
      props: {
        base: false,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should add decorator class with left/right props', async () => {
    const wrapper = mountFunction({
      props: {
        left: true,
      },
    })

    expect(wrapper.classes('icon--left')).toBe(true)

    await wrapper.setProps({ right: true })

    expect(wrapper.classes('icon--right')).toBe(true)
  })

  it('should override "aria-hidden" from attributes', async () => {
    const wrapper = mountFunction({
      attrs: {
        'aria-hidden': true,
      },
    })

    expect(wrapper.attributes('aria-hidden')).toBe('true')
  })
})
