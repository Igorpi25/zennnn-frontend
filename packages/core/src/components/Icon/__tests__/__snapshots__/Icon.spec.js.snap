// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon.js should have correct size style on size prop and match snapshot 1`] = `
<i class="icon" aria-hidden="true"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" height="20px" width="20px" role="img" aria-hidden="true" style="font-size: 20px; height: 20px; width: 20px;">
    <g fill="currentColor">
      <path d="M18 12a6 6 0 11-12 0 6 6 0 0112 0z"></path>
    </g>
  </svg></i>
`;

exports[`Icon.js should not set size style on base prop false and match snapshot 1`] = `
<i class="icon" aria-hidden="true"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" height="24" width="24" role="img" aria-hidden="true">
    <g fill="currentColor">
      <path d="M18 12a6 6 0 11-12 0 6 6 0 0112 0z"></path>
    </g>
  </svg></i>
`;

exports[`Icon.js should render "button" tag on attrs.onClick exists and match snapshot 1`] = `
<button class="icon icon--link" role="button" aria-hidden="false"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" height="24px" width="24px" role="img" aria-hidden="true" style="font-size: 24px; height: 24px; width: 24px;">
    <g fill="currentColor">
      <path d="M18 12a6 6 0 11-12 0 6 6 0 0112 0z"></path>
    </g>
  </svg></button>
`;

exports[`Icon.js should render component and match snapshot 1`] = `
<i class="icon" aria-hidden="true"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" height="24px" width="24px" role="img" aria-hidden="true" style="font-size: 24px; height: 24px; width: 24px;">
    <g fill="currentColor">
      <path d="M18 12a6 6 0 11-12 0 6 6 0 0112 0z"></path>
    </g>
  </svg></i>
`;

exports[`Icon.js should render specified tag when using tag prop and match snapshot 1`] = `<span class="icon" aria-hidden="true"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" height="24px" width="24px" role="img" aria-hidden="true" style="font-size: 24px; height: 24px; width: 24px;"><g fill="currentColor"><path d="M18 12a6 6 0 11-12 0 6 6 0 0112 0z"></path></g></svg></span>`;

exports[`Icon.js should set size from helper prop and match snapshot 1`] = `
<i class="icon" aria-hidden="true"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" height="18px" width="18px" role="img" aria-hidden="true" style="font-size: 18px; height: 18px; width: 18px;">
    <g fill="currentColor">
      <path d="M18 12a6 6 0 11-12 0 6 6 0 0112 0z"></path>
    </g>
  </svg></i>
`;

exports[`Icon.js should set size from helper prop and match snapshot 2`] = `
<i class="icon" aria-hidden="true"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" height="24px" width="24px" role="img" aria-hidden="true" style="font-size: 24px; height: 24px; width: 24px;">
    <g fill="currentColor">
      <path d="M18 12a6 6 0 11-12 0 6 6 0 0112 0z"></path>
    </g>
  </svg></i>
`;

exports[`Icon.js should set size from helper prop and match snapshot 3`] = `
<i class="icon" aria-hidden="true"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" height="32px" width="32px" role="img" aria-hidden="true" style="font-size: 32px; height: 32px; width: 32px;">
    <g fill="currentColor">
      <path d="M18 12a6 6 0 11-12 0 6 6 0 0112 0z"></path>
    </g>
  </svg></i>
`;
